### ISSUE TITLE
Weighted least squares

### ISSUE THREAD

**Issue by fritschf** [fabiodimarco/torch-levenberg-marquardt#3]
Hi @fabiodimarco!
Thank you for versatile Pytorch implementation of the LM algorithm. 
I wonder if it would be possible to extend the library to include weighted least squares. 
I suspect that special care needs to be taken here regarding the mini-batching.

**Comment by fabiodimarco** [comment]
Hi @fritschf, sure the weights can be added to the loss easily. I suggest you take a look at the `loss.py` file where the losses are defined.
What do you want to weight the outputs or the batch examples?

**Comment by fritschf** [comment]
Thank you for your quick response. I want to weight every output individually to highlight the importance of some data points. 
My first naive attempt was initializing the loss function with a tensor containing the weights for all data points, but this attempt failed due to the mini-batching.

**Comment by fabiodimarco** [comment]
@fritschf why is mini-batching a problem? I think you can define your own loss function that satisfies your needs.

Is you want to use MSELoss with weights you can do something like that:

```
class MSELoss(tlm.loss.Loss):
    def __init__(self, weights: Tensor) -> None:
        super().__init__()
        self.register_buffer("weights", weights)

    def forward(self, y_true: Tensor, y_pred: Tensor) -> Tensor:
        w = self.weights.view(1, -1, *[1] * (y_true.ndim - 2))
        return ((w * (y_pred - y_true)).square()).mean()

    def residuals(self, y_true: Tensor, y_pred: Tensor) -> Tensor:
        w = self.weights.view(1, -1, *[1] * (y_true.ndim - 2))
        return w * (y_pred - y_true)
```

Of course the number of weights must be equal to the number of outputs. Let me know if that is want you are looking for.
I could add the weight parameter to the default MSELoss implementation.

**Comment by fritschf** [comment]
Your implementation seems to work. Thank you! 
However,  with this new loss function the runtime of the otherwise unchanged program increased drastically. I'll take a look myself to see what the cause is.

**Comment by fabiodimarco** [comment]
@fritschf maybe you are running on CPU instead of GPU? In that case you have to move the instance of the MSELoss into the correct device.

**Comment by fritschf** [comment]
Thanks for the hint, I will look into it.

**Comment by fritschf** [comment]
Ok, the CPU/GPU seems to not be the issue. 
However, I suspect there still seems to be an issue about the dimensions of the tensor:
The weight tensor does not have the same size as the y_pred and y_true tensors. Its leading dimension is of the size of the full dataset instead of the size of one mini-batch. 

To avoid any misunderstanding of my initial goal: the output should be weighted individually for each sample

Maybe you can help me out with that again.

**Comment by fabiodimarco** [comment]
@fritschf can you share your code so that I can run it?

**Comment by fritschf** [comment]
Sure, to simplify things I have modified your example where I also see the increased runtime. 

```python

import time

import torch
import torch_levenberg_marquardt as tlm
from torch.utils.data import TensorDataset


class WeightedMSELoss(tlm.loss.Loss):
    def __init__(self, weights: torch.Tensor) -> None:
        super().__init__()
        self.register_buffer("weights", weights)

    def forward(self, y_true: torch.Tensor, y_pred: torch.Tensor) -> torch.Tensor:
        w = self.weights.view(1, -1, *[1] * (y_true.ndim - 2))
        return ((w * (y_pred - y_true)).square()).mean()

    def residuals(self, y_true: torch.Tensor, y_pred: torch.Tensor) -> torch.Tensor:
        w = self.weights.view(1, -1, *[1] * (y_true.ndim - 2))
        return w * (y_pred - y_true)



# Set PyTorch to use high precision for matrix multiplication
torch.set_float32_matmul_precision('high')
torch.set_default_dtype(torch.float64)
torch.manual_seed(0)

# Detect CUDA device for acceleration
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f'Using device: {device}')


# %%
# Generate synthetic dataset for training
input_size = 2000  # Total number of data points
batch_size = 100  # Number of samples per batch

# Generate training inputs and outputs (y = sinc(10 * x))
x_train = torch.linspace(-1, 1, input_size, dtype=torch.float64).unsqueeze(1).to(device)
y_train = ( torch.cos(10 * x_train) * torch.sin( x_train ** 2) ).to(device)

#Generate weights for each input
weights = torch.exp(-2.0 * abs(y_train)).to(device)

# Create dataset and dataloader
train_dataset = TensorDataset(x_train, y_train)
train_loader = tlm.utils.FastDataLoader(
    train_dataset,
    batch_size=batch_size,
    repeat=1,
    shuffle=True,
    device=device,
)


# %%
# Define a function to create the neural network model
def create_model() -> torch.nn.Module:
    return torch.nn.Sequential(
        torch.nn.Linear(1, 40),
        torch.nn.Tanh(),
        torch.nn.Linear(40, 1),
    ).to(device)


# Initialize models for Adam and Levenberg-Marquardt optimization
model_lm = create_model()

# Print the number of trainable parameters
num_parameters = sum(p.numel() for p in model_lm.parameters() if p.requires_grad)
print(f'Number of trainable parameters: {num_parameters}')


module_lm = tlm.training.LevenbergMarquardtModule(
    model=model_lm,
#    loss_fn=tlm.loss.MSELoss(),
    loss_fn=WeightedMSELoss(weights),
    learning_rate=1.0,
    attempts_per_step=10,
    solve_method='qr',
)

# Train the model using the Levenberg-Marquardt algorithm
print('Training with Levenberg-Marquardt...')
t2_start = time.perf_counter()

tlm.utils.fit(
    module_lm,
    train_loader,
    epochs=10,
)

t2_stop = time.perf_counter()
print(f'Training completed. Elapsed time: {t2_stop - t2_start:.2f} seconds')

**Comment by fabiodimarco** [comment]
@fritschf Ok, I miss understood what you wanted to achieve. The previous loss then is wrong, as it was meant to introduce a different weight for each output for a model with more then one output. While what you want to achieve is a different weight for each training example. Here is a way you can achieve that:

```
import time

import torch
import torch_levenberg_marquardt as tlm
from torch import Tensor
from torch.utils.data import TensorDataset


class WeightedMSELoss(tlm.loss.Loss):
    def __init__(self) -> None:
        super().__init__()

    def forward(self, y_true: Tensor, y_pred: Tensor) -> Tensor:
        w = y_true[..., 1]
        y_true = y_true[..., 0]
        return ((w * (y_pred - y_true)).square()).mean()

    def residuals(self, y_true: Tensor, y_pred: Tensor) -> Tensor:
        w = y_true[..., 1]
        y_true = y_true[..., 0]
        return w * (y_pred - y_true)


# Set PyTorch to use high precision for matrix multiplication
torch.set_float32_matmul_precision('high')
torch.set_default_dtype(torch.float64)
torch.manual_seed(0)

# Detect CUDA device for acceleration
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f'Using device: {device}')


# %%
# Generate synthetic dataset for training
input_size = 2000  # Total number of data points
batch_size = 100  # Number of samples per batch

# Generate training inputs and outputs (y = sinc(10 * x))
x_train = torch.linspace(-1, 1, input_size, dtype=torch.float64).unsqueeze(1).to(device)
y_train = (torch.cos(10 * x_train) * torch.sin(x_train**2)).to(device)

# Generate weights for each input
weights = torch.exp(-2.0 * abs(y_train)).to(device)

targets = torch.stack([y_train, weights], dim=-1)

# Create dataset and dataloader
train_dataset = TensorDataset(x_train, targets)
train_loader = tlm.utils.FastDataLoader(
    train_dataset,
    batch_size=batch_size,
    repeat=1,
    shuffle=True,
    device=device,
)


# %%
# Define a function to create the neural network model
def create_model() -> torch.nn.Module:
    return torch.nn.Sequential(
        torch.nn.Linear(1, 40),
        torch.nn.Tanh(),
        torch.nn.Linear(40, 1),
    ).to(device)


# Initialize models for Adam and Levenberg-Marquardt optimization
model_lm = create_model()

# Print the number of trainable parameters
num_parameters = sum(p.numel() for p in model_lm.parameters() if p.requires_grad)
print(f'Number of trainable parameters: {num_parameters}')


module_lm = tlm.training.LevenbergMarquardtModule(
    model=model_lm,
    # loss_fn=tlm.loss.MSELoss(),
    loss_fn=WeightedMSELoss(),
    learning_rate=1.0,
    attempts_per_step=10,
    solve_method='qr',
)

# Train the model using the Levenberg-Marquardt algorithm
print('Training with Levenberg-Marquardt...')
t2_start = time.perf_counter()

tlm.utils.fit(
    module_lm,
    train_loader,
    epochs=50,
)

t2_stop = time.perf_counter()
print(f'Training completed. Elapsed time: {t2_stop - t2_start:.2f} seconds')
```

Let me know if that is what you wanted.

**Comment by fritschf** [comment]
That's it, easier then expected. Thanks again for your help!
